website/server/libs/payments.js

```diff
@@ -321,4 +322,180 @@ api.buyGems = async function buyGems (data) {
   await data.user.save();
 };
 
+/**
+ * Allows for purchasing a user subscription, group subscription or gems with Stripe
+ *
+ * @param  options
+ * @param  options.token  The stripe token generated on the front end
+ * @param  options.user  The user object who is purchasing
+ * @param  options.gift  The gift details if any
+ * @param  options.sub  The subscription data to purchase
+ * @param  options.groupId  The id of the group purchasing a subscription
+ * @param  options.email  The email enter by the user on the Stripe form
+ * @param  options.headers  The request headers to store on analytics
+ * @return undefined
+ */
+api.payWithStripe = async function payWithStripe (options, stripeInc) {
+  let [
+    token,
+    user,
+    gift,
+    sub,
+    groupId,
+    email,
+    headers,
+  ] = options;
+  let coupon;
+  let response;
+  let subscriptionId;
+  // @TODO: We need to mock this, but curently we don't have correct Dependency Injection
+  let stripeApi = stripe;
+
+  if (stripeInc) stripeApi = stripeInc;
+
+  if (!token) throw new BadRequest('Missing req.body.id');
+
+  if (sub) {
+    if (sub.discount) {
+      if (!req.query.coupon) throw new BadRequest(res.t('couponCodeRequired'));
+      coupon = await Coupon.findOne({_id: cc.validate(req.query.coupon), event: sub.key});
```

Comment:
otherwise it's not a full promise, see http://mongoosejs.com/docs/promises.html under `Queries are not promises`